compiler/src/iree/compiler/Dialect/Encoding/IR/EncodingOps.h

```diff
@@ -42,6 +42,17 @@ EncodingAttr getEncodingAttr(RankedTensorType type);
 /// Returns the ContractionDimensions for the encoding user_indexing_maps.
 FailureOr<linalg::ContractionDimensions>
 getEncodingContractionDims(EncodingAttr encoding);
+
+/// Named enum of the encoding role
+// Assign a name to the index for clarity
+enum RoleIndex : int64_t {
+  MATMUL_LHS = 0,
+  MATMUL_RHS = 1,
+  MATMUL_RESULT = 2,
+  // TODO: conv
+};
```

Comment:
We rarely introduce unscoped enum declarations like this --- we tend to only use `enum class`, which provides type safety, for cases where an enum is wanted. Here, though, there is clearly a reason why this doesn't use `enum class`: this has to interoperate with this being a plain integer in the updated `.td`. Making this an enum class would be inconvenient indeed (would need to `static_cast` everywhere).

The problem with unscoped enums like this is that they really are just integer constants, despite looking like enums, so they can be misleading: someone might think that the compiler is enforcing some typing here, when it's really not.

If constants like this were needed, I would actually rather have them be explicitly raw integer constants, to make it clearer that that is what these are:

`static constexpr int MATMUL_RESULT = 2`.

However, I don't think that this is really needed. This is not specific at all to encodings: a constant like `static constexpr int MATMUL_RESULT = 2` would be a general fact about `linalg.matmul`. Adding it is amending a problem that preexists with `linalg.matmul` and Linalg in general, that one has to refer to operands by integral indices that act as magic constants.

There is no need for us to try to address that general Linalg issue. I would just not have any symbolic constant and have the code use hardcoded literals 0, 1, 2. This is already what existing code is doing. In specific contexts you will find opportunities to introduce some helpful local naming and/or commenting.

Reply:
Updated